src/storage/store/engine.rs

```diff
@@ -861,55 +865,173 @@ impl ShardEngine {
             })
         });
 
-        for msg in &sorted_user_messages {
-            // Errors are validated based on the shard root
-            match self.validate_user_message(msg, timestamp, version, txn_batch) {
-                Ok(()) => {
-                    let result = self.merge_message(msg, txn_batch);
-                    match result {
-                        Ok(event) => {
-                            merged_messages_count += 1;
-                            self.update_trie(trie_ctx, &event, txn_batch)?;
-                            events.push(event.clone());
-                            user_messages_count += 1;
-                            message_types.insert(msg.msg_type());
+        if source == ProposalSource::Replication {
```

Comment:
Hmm, if we're not doing the rest of the validation, do we even gain much from using the engine logic? Should we just move this out of the engine? Or at least out of `replay_snapchain_txn`? I think it might be cleaner if we don't try to make this fit into existing engine logic.